### DISCUSSION TITLE
Newsletter email subscribers

### DISCUSSION THREAD

**Discussion by meysam81** [squidfunk/mkdocs-material#7104] [Q&A] [answered]
Hey, great people,

It goes without saying that you're doing an awesome job.

I wonder if there is anyway we can grow a newsletter and asking people for their emails; one which will be used to push new posts right into their inbox.

I have seen [Adding buttons](https://squidfunk.github.io/mkdocs-material/reference/buttons/?h=subscrib#adding-buttons).

I have also seen [this discussion](https://github.com/squidfunk/mkdocs-material/discussions/6099) in this repository.

Yet, I am not sure how this functionality works. Do you, e.g., use ConvertKit or something similar to get their emails? And what happens after that? Do you `mkdocs build` the docs and use [offline plugin](https://squidfunk.github.io/mkdocs-material/plugins/offline/?h=offline#built-in-offline-plugin) to send one HTML file to their inbox? Do you manipulate the DOM in a customized way? Would you say that Mailchimp does this out of the box?

I'm a bit unclear as to how to grow a newsletter with email subscribers the same way you'd have in platforms like Medium or Substack.

Much appreciated in advance.

**Comment by kamilkrzyskow** [comment]
Hello @meysam81,
I don't have experience with neither growing a mailing list nor using Medium/Substack, but I might be able to point you in the right direction or at least start the discussion 😅

1. As far, as I know Material for MkDocs has no newsletter and the buttons in the docs are simply example of generic call to actions.
    I also think that providing more explicit instructions on how to grow a mailing list is outside of the project's scope. However, I can see your point that people might migrate from other blogging platforms here, and might expect a swap-in solution for newsletters.
2. In the [discussion](https://github.com/squidfunk/mkdocs-material/discussions/6099) you mentioned Tim Vink talks about his MailChimp integration, we can see the rendered blog and the source code here:

    - https://timvink.nl/blog/
    - https://github.com/timvink/personal-site/blob/main/docs/blog/index.md?plain=1

    We can see that's a HTML form which sends data to a specific web address, he got (I'm assuming) from his MailChimp account dashboard `form action="https://timvink.us20.list-manage.com/subscribe/post?...`, so that's how he gets the emails. 
    If you want to get more info about what happens next, just ping Tim in this thread. You likely can schedule weekly posts that will be sent to the list. 
    Based on his call to action "Get notified of new posts" I would assume his list sends a generic message like "Tim Vink posted a new article titled ... -> go see the website <-" and funnels the readers back to the website.
    > Would you say that Mailchimp does this out of the box?
    
    Your idea might also be correct, as web scraping isn't that complicated, so they could offer an option to scrape your own site and send it as an email without you having to crank the knobs.
3. I would expect the mailing list manager service to provide some templates for your newsletter. However, it would likely be best to match the look of the website, so it also has to accept some sort\* of HTML files as your content. The issue is that the offline plugin likely won't help here, as it only changes some things to make sure that the website is "_functional_" when running from the filesystem. It still has relative paths to CSS/JavaScript assets, pages still contain the navigation tree, and it still loads the search bar. 
    \*So I think you would need to define a custom template file, similar to the `home.html` override template, but for an article that will be posted to your newsletter. This template would need to force absolute direct URLs to your assets, remove navigation items and unnecessary JavaScript, it should be a plain HTML page styled same as your actual blog page. Just copy the [`base.html`](https://github.com/squidfunk/mkdocs-material/blob/master/src/templates/base.html) and modify it until you get a satisfactory result.
    
\*I can imagine some sort of security/optimization measure that doesn't allow customization of the email HTML structure too much, and only allows to style the structure provided by the mailing list manager, so best to check the capabilities before diving into custom template creation.

---

Sorry for the lack of specifics, and I hope my rambling helped you in some way.

**Reply by timvink** [reply, marked as answer]
Hello, Tim here :)

Setting up email subscriptions to a static site is something completely separate from mkdocs. There are two components basically:

1. Collecting email addresses. I used a form supplied by Mailchimp, and improved the layout a bit using HTML/CSS
2. Triggering email campaigns to your subscribers. I setup Mailchimp to check every night my website's RSS feed, and send out the newest post as an email.

To get the RSS feed from my mkdocs-material blog, I used the mkdocs RSS plugin.

Good luck!

**Comment by meysam81** [comment]
@squidfunk 

I've been trying different things related to the topic of this discussion.

Based on my understanding, the best way to achieve a newsletter-like with mkdocs blogs is probably some manual manipulation of the DOM to grab the content of the article and pass that as the body of an email-template to the subscribers (whichever platform it is).

Would you say there is something wrong with the following approach? Should I not rely on these class and element names? Is it that just pinning my `requirements.txt` to a specific version can keep my content safe from breaking? Or is there more to it than that?

```js
document.querySelector('header[data-md-component="header"]').remove()
document.querySelector('div[data-md-component="announce"]').remove()
document.querySelector('div.md-sidebar.md-sidebar--post').remove()
document.querySelector('div.md-sidebar.md-sidebar--primary').remove()
document.querySelector('div.md-sidebar.md-sidebar--secondary').remove()
document.querySelector('aside.md-source-file').remove()
document.querySelector('form.md-feedback').remove()
document.querySelector('footer.md-footer').remove()

document.querySelectorAll('body > :not(div[data-md-component="container"])').forEach(e => e.remove())
```

**Reply by squidfunk** [reply]
I'm not sure I understand what exactly you want to achieve, but the approach above will only work with JavaScript. It will not work when users have disabled JavaScript, and/or are on slow connections.

**Reply by kamilkrzyskow** [reply]
I'm also unsure why are you using DOM manipulation via JavaScript. If you're going to change the structure of the website, best to do it via a custom template and theme extension as I mentioned in my 3rd point above. I'm not sure if emails can contain JavaScript, but best send plain HTML.
https://squidfunk.github.io/mkdocs-material/reference/#setting-the-page-template
https://squidfunk.github.io/mkdocs-material/customization/#extending-the-theme

To answer other questions, yes, if you lock the version of the theme in the `requirements.txt` then the DOM structure will not change, so targetting the selectors should always work. AFAIK only major versions modify the HTML structure so 8.x, 9.x etc., so you can lock the version to <10

EDIT: There is also the option to extract only the HTML content of a page via some hook and the on_post_page event:
https://www.mkdocs.org/user-guide/configuration/#hooks
https://www.mkdocs.org/dev-guide/plugins/#on_post_page

**Reply by meysam81** [reply]
Well, I don't want to change the page as it is, i.e., if anyone is visiting the website, I want them to see the same thing that Material for Mkdocs is rendering from my Markdown files. That's that, and the content is good (perfect, really).

The requirement I am addressing at this point is to somehow, some way, grab the latest posts (e.g., using RSS plugin), and send the content to the subscribers of my newsletter. I don't know how much you've been around Medium, Substack, Beehiv, Hashinode, etc.
The idea is that my readers provide me with their emails, and I will send the latest blog post as soon as it is published.

The best way I've come up with so far is this ugly DOM manipulation; imagine a JS code running somewhere on the server that fetches my latest blog post, modifies the DOM, and sends the content of the post without all the headers, sidebars and everything else that's not in the post itself.

**Reply by timvink** [reply]
I think you're probably overcomplicating things.

Enable this plugin so your website has an RSS feed: https://guts.github.io/mkdocs-rss-plugin/

Then use a newsletter provider (I use the free version of Mailchimp) and point it to the RSS url of your website. It can check daily or weekly for new content, and then send that your subscribers.

HTML email formatting is notoriously tricky. Forget about trying to get the exact same styles and use the email template from your newsletter provider. The content from the RSS feed is inserted. So the RSS plugin already takes care of extracting the body of your blogpost, no need to mess around with the DOM.

**Reply by meysam81** [reply]
There is a catch, however. The RSS plugin, even when receiving the entire content of the post<sup>[1](https://guts.github.io/mkdocs-rss-plugin/configuration/#abstract_delimiter-abstract-delimiter)</sup>, is going to email the markdown content of your files.

This means, all those [admonitions](https://squidfunk.github.io/mkdocs-material/reference/admonitions/), [content tabs](https://squidfunk.github.io/mkdocs-material/reference/content-tabs/?h=tab#content-tabs), and all the other beautiful elements of Material for Mkdocs are not sent to their inbox as rendered on the website, but fetched from the markdwon file and sent as plain & ugly content.

That is no different than sending the plain `.md` file; and that can't be worse really!

**Reply by timvink** [reply]
I see an opportunity for someone to build a newsletter provider that supports the full range python-markdown & extensions :) https://facelessuser.github.io/pymdown-extensions/

A more basic solution is to just send an email with the title, first paragraph and a link to your blogpost

**Reply by meysam81** [reply]
This second part was what I was aiming for with Mailchimp at the beginning, having realized that the RSS plugin is not sending the rendered HTML but the markdown file.

But then it occurred to me, "_if they subscribed to the newsletter, why should they bother open the link and come to the website?_", like, "_isn't the whole point of subscribing that you will get the content?_"

I do not mean to trick the readers/subscribers, and as such, this doesn't look like a sexy approach.

+1 to the opportunity and a heroic opportunist. :muscle:

**Reply by kamilkrzyskow** [reply]
After having a quick look at some emails in my inbox, I confirmed 3 things:
- All use inline CSS, some include a `<style>` tag in the `<head>`, but no external CSS, which matches the info at:
    https://mailtrap.io/blog/email-css/
- All links to images are absolute, so the default behaviour of relative links in MkDocs would have to be modified, I remember there were some similar requests for this feature in the past 🤔 
- For the most part the positioning of items is configured via HTML tables, even when `div` and `p` tags are all around, at least one global table is present

So converting a Material for MkDocs page, would entail extracting the external base CSS into either the `<style>` tag or for better compatibility with various email clients into inline CSS for each element and also it might require converting the layout into a table 🤔 

IMO too complicated to be worth the hassle, as each each modification to the HTML via some markdown extension could possibly ruin the rest 😅 Unless there would be specific rules set in place, that a blog post mustn't use XYZ features to make it compatible with email, but that's less than ideal 🤔

**Comment by AtticusZeller** [comment]
[good example](https://madewithml.com/#mlops) here but I don't  how he realized it, there no much information in it's repo
![IMG_20241117_111858.jpg](https://github.com/user-attachments/assets/b468b38d-56eb-4671-86ed-922d17c3e1e7)

**Reply by kamilkrzyskow** [reply]
The site uses the `material` theme, while the `mkdocs.yml` shows `readthedocs`, so the repository hasn't been updated with the latest files.

As for the "Subscribe" button opens a Bootstrap modal, and in that modal there is an iframe with `sibforms.com` link, which seems to refer to [Brevo Forms](https://www.brevo.com/).

There is also this odd comment:
```
DON'T FORGET TO CHANGE LINK INSIDE docs/overrides/newsletter.html and mkdocs.yml 
```
which I wouldn't make for myself personally, so perhaps this is a public template, available somewhere.
However, with the current state of search engines, I could not find anything 🤷 

Anyway, the comment implies usage of template overrides with the required form embed:
- https://squidfunk.github.io/mkdocs-material/customization/#extending-the-theme

**Reply by AtticusZeller** [reply]
thank you for such professional explanation 😊@kamilkrzyskow
